packages/react/src/components/createInlineOverlayComponent.tsx

```diff
@@ -107,13 +114,38 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
           this.setState({ isOpen: false });
         }
 
-        this.props.onDidDismiss && this.props.onDidDismiss(evt);
+        const didComponentUnmount = this.ref.current === null;
+
+        if (didComponentUnmount === false) {
+          /**
+           * If the component is unmounted, we do not
+           * want to call the onDidDismiss handler to
+           * prevent a React state update on an unmounted component.
+           */
+          this.props.onDidDismiss && this.props.onDidDismiss(evt);
+        }
       });
     }
 
     componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
       const node = this.ref.current! as HTMLElement;
-      attachProps(node, this.props, prevProps);
+      const propsToPass = Object.keys(this.props).reduce((acc, name) => {
+        if (!excludedOverlayEvents.includes(name)) {
+          /**
+           * Maintains all props except for the event handlers
+           * that are manually handled in the componentDidMount.
+           */
+          (acc as any)[name] = (this.props as any)[name];
+        }
+        return acc;
+      }, {});
+      attachProps(node, propsToPass, prevProps);
+    }
+
+    componentWillUnmount() {
+      if (this.ref.current) {
+        this.ref.current.dismiss();
+      }
```

Comment:
If React handles the rendering of the Web Component, won't the `ion-modal` element be removed automatically?

Reply:
React will only remove the DOM node if the overlay is dismissed. When the modal is presented, we are manually moving the node to the root and React is unable to know to remove it when the React component unmounts. This check/code moves the node back in time for React to remove it.